atomic_reactor/plugins/build_orchestrate_build.py

```diff
@@ -0,0 +1,224 @@
+
+    def wait_to_finish(self):
+        self.build = self.osbs.wait_for_build_to_finish(self.name)
+        return self.build
+
+    def watch_logs(self, logger):
+        for line in self.osbs.get_build_logs(self.name, follow=True):
+            # TODO: This is a little clunky:
+            # 2017-02-24 14:22:51,472 - atomic_reactor.plugins.orchestrate_build - INFO - x86_64 - 2017-02-24 14:22:51,314 - atomic_reactor.plugin - INFO - <worker build log message>
+            logger.info('%s - %s', self.platform, line)
+
+    def get_annotations(self):
+        build_annotations = self.build.get_annotations() or {}
+        return {
+            'build': {
+                'cluster-url': self.osbs.os_conf.get_openshift_base_uri(),
+                'namespace': self.osbs.os_conf.get_namespace(),
+                'build-name': self.name,
+            },
+            'digests': json.loads(
+                build_annotations.get('digests', '[]')),
+            'plugins-metadata': json.loads(
+                build_annotations.get('plugins-metadata', '{}')),
+        }
+
+    def get_fail_reason(self):
+        if not self.build:
+            return {'general': 'build not started'}
+        metadata = json.loads(
+                self.build.get_annotations().get('plugins-metadata', '{}'))
+        return metadata.get('errors', {})
+
+    def cancel_build(self):
+        if self.build and not self.build.is_finished():
+            self.osbs.cancel_build(self.name)
+
+
+class OrchestrateBuildPlugin(BuildStepPlugin):
+
+    EXCLUDE_FILENAME = 'exclude-platform'
+
+    key = 'orchestrate_build'
+
+    def __init__(self, tasker, workflow, platforms, build_kwargs):
+        """
+        constructor
+
+        :param tasker: DockerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param platforms: list<str>, platforms to build
+        :param build_kwargs: dict, keyword arguments for starting worker builds
+        """
+        super(OrchestrateBuildPlugin, self).__init__(tasker, workflow)
+        self.platforms = set(platforms)
+        self.build_kwargs = build_kwargs
+
+        self.worker_builds = []
+
+    def get_excluded_platforms(self):
+        df_dir = self.workflow.source.get_dockerfile_path()[1]
+        exclude_platforms = set()
+        exclude_path = os.path.join(df_dir, self.EXCLUDE_FILENAME)
+        if os.path.exists(exclude_path):
+            with open(exclude_path) as f:
+                for platform in f:
+                    platform = platform.strip()
+                    if not platform:
+                        continue
+                    exclude_platforms.add(platform)
+
+        return exclude_platforms
+
+    def get_platforms(self):
+        return self.platforms - self.get_excluded_platforms()
+
+    def get_current_builds(self, osbs):
+        field_selector = ','.join(['status!={status}'.format(status=status.capitalize())
+                                   for status in BUILD_FINISHED_STATES])
```

Comment:
We could apply some labels when creating the worker build?

I'm thinking now that this is actually needed for using a cluster for both orchestrator and worker builds. If we count all current builds together, the computed load will, likely, always be higher in the orchestrator cluster, which means it won't be chosen! This, of course, depends on the `max_concurrent_builds` setting for each cluster.

Having said that, it seems beyond the scope of this change. We need to have changes in osbs-client first in order to leverage it in this plugin. How about we focus this PR on adding ability to orchestrate builds, and create a new story for allowing orchestrator cluster to also be used for worker builds?